extensions/positron-assistant/src/tools/projectTreeTool.ts

```diff
@@ -0,0 +1,287 @@
+
+		// Parse the input options
+		const { include, exclude, replaceDefaultExcludes, excludeSettings, ignoreFiles, filterResults, maxFiles } = options.input;
+
+		log.trace(`[${PositronAssistantToolName.ProjectTree}] Invoked with options: ${JSON.stringify(options.input, null, 2)}`);
+
+		const filePatterns = include ?? DEFAULT_INCLUDE_PATTERNS;
+		const excludePatterns = exclude ?? [];
+		const filterResultsEnabled = filterResults ?? DEFAULT_FILTER_RESULTS;
+		const filesLimit = maxFiles ?? DEFAULT_MAX_FILES;
+
+		let findOptions: vscode.FindFiles2Options;
+		if (filterResultsEnabled) {
+			findOptions = {
+				exclude: replaceDefaultExcludes ? excludePatterns : [...DEFAULT_EXCLUDE_PATTERNS, ...excludePatterns],
+				useIgnoreFiles: ignoreFiles === false ? undefined : DEFAULT_USE_IGNORE_FILES,
+				useExcludeSettings: excludeSettings ? getExcludeSettingOptions(excludeSettings) : DEFAULT_EXCLUDE_SETTING_OPTIONS,
+			};
+		} else {
+			findOptions = {
+				exclude: undefined,
+				useIgnoreFiles: {
+					local: false,
+					parent: false,
+					global: false,
+				},
+				useExcludeSettings: vscode.ExcludeSettingOptions.None,
+			};
+		}
+
+
+		log.trace(`[${PositronAssistantToolName.ProjectTree}] Constructing project tree with options: ${JSON.stringify({
+			include: filePatterns,
+			exclude: findOptions.exclude,
+			useIgnoreFiles: findOptions.useIgnoreFiles,
+			useExcludeSettings: findOptions.useExcludeSettings,
+			filterResults: filterResultsEnabled,
+			maxFiles: filesLimit,
+		}, null, 2)}`);
+
+		// Construct the project tree
+		const workspaceTrees: DirectoryInfo[] = [];
+		for (const folder of workspaceFolders) {
+			// NOTE: this will not include empty directories :/
+			const matchedFileUris = await vscode.workspace.findFiles2(
+				filePatterns,
+				findOptions,
+				token
+			);
+			const items = convertUrisToDirectoryItems(folder, matchedFileUris);
+			workspaceTrees.push({ folder, items, totalFiles: matchedFileUris.length });
+		}
+
+		const totalFiles = workspaceTrees.reduce((sum, obj) => sum + obj.totalFiles, 0);
+
+		log.debug(`[${PositronAssistantToolName.ProjectTree}] Project tree constructed with ${totalFiles} items across ${workspaceFolders.length} workspace folders.`);
+
+		// Return a compressed description of the project tree if there are too many items
+		if (totalFiles > filesLimit) {
+			const itemLimit = Math.floor(filesLimit / workspaceTrees.length);
+			log.debug(`[${PositronAssistantToolName.ProjectTree}] Project tree exceeds the limit of ${filesLimit} items. A summary will be returned for each workspace folder.`);
+			const summarizedTree = await getSummarizedProjectTree(workspaceTrees, itemLimit);
+			return new vscode.LanguageModelToolResult([
+				new vscode.LanguageModelTextPart(`Project tree contains ${totalFiles} files, which exceeds the limit of ${filesLimit}. Here is a summary of each workspace, including the first ${itemLimit} files and directories:`),
```

Comment:
note: `totalFiles` is a bit awkward, because it is not necessarily the _real_ total files that you would get with running e.g. `find . -type f | wc -l` in the directory. It's the total files that were found based on the options passed to `findFiles2()`, which may be excluding a number of files.

We may want a separate tool/participant/etc that provides general workspace metadata, like file count, open editors, workspace folders, unstaged files, etc.